From: Chad Dupuis <cdupuis@redhat.com>
Date: Tue, 30 Nov 2010 19:58:59 -0500
Subject: [scsi] qla4xxx: no wait for outstanding command complete
Message-id: <20101130195859.10450.39330.sendpatchset@localhost.localdomain>
Patchwork-id: 29693
O-Subject: [RHEL 5.6 PATCH 2/28] qla4xxx: Do not wait for the outstanding
	commands to complete
Bugzilla: 656999

Bugzilla
--------

Bug 656999 (https://bugzilla.redhat.com/show_bug.cgi?id=656999)

Upstream Status
---------------

Upstream posting: http://marc.info/?l=linux-scsi&m=128809710800356&w=2

Description
-----------

>From 80d739ac191dc7b9a75246875a804e332bcdead5 Mon Sep 17 00:00:00 2001
From: Nilesh Javali <nilesh.javali@qlogic.com>
Date: Sat, 20 Nov 2010 14:00:15 -0800
Subject: [PATCH 03/29] qla4xxx: Do not wait for the outstanding commands to complete

Do not wait for the outstanding commands to complete in case
of firmware hang.

Signed-off-by: Nilesh Javali <nilesh.javali@qlogic.com>
Signed-off-by: Prasanna Mumbai <prasanna.mumbai@qlogic.com>

diff --git a/drivers/scsi/qla4xxx/ql4_os.c b/drivers/scsi/qla4xxx/ql4_os.c
index a771145..1d899da 100644
--- a/drivers/scsi/qla4xxx/ql4_os.c
+++ b/drivers/scsi/qla4xxx/ql4_os.c
@@ -1153,7 +1153,8 @@ static int qla4xxx_recover_adapter(struct scsi_qla_host *ha,
 				ha->host_no, __func__));
 			status = ha->isp_ops->reset_firmware(ha);
 			if (status == QLA_SUCCESS) {
-				(void) qla4xxx_cmd_wait(ha);
+				if (!test_bit(AF_FW_RECOVERY, &ha->flags))
+					qla4xxx_cmd_wait(ha);
 				ha->isp_ops->disable_intrs(ha);
 				qla4xxx_process_aen(ha, FLUSH_DDB_CHANGED_AENS);
 				qla4xxx_abort_active_cmds(ha, DID_RESET << 16);
@@ -1170,7 +1171,8 @@ static int qla4xxx_recover_adapter(struct scsi_qla_host *ha,
 	 * or if stop_firmware fails for ISP-802x.
 	 * This is the default case for ISP-4xxx */
 	if (!is_qla8022(ha) || reset_chip) {
-		(void) qla4xxx_cmd_wait(ha);
+		if (!test_bit(AF_FW_RECOVERY, &ha->flags))
+			qla4xxx_cmd_wait(ha);
 		qla4xxx_process_aen(ha, FLUSH_DDB_CHANGED_AENS);
 		qla4xxx_abort_active_cmds(ha, DID_RESET << 16);
 		DEBUG2(dev_info(&ha->pdev->dev,
